### ISSUE TITLE
Intercepting JS/WASM exceptions

### ISSUE THREAD

**Issue by tklajnscek** [emscripten-core/emscripten#11667]
_Note: This post is not about C++ exceptions!_

If a program aborts (e.g. out-of-memory and similar) there's the Module.onAbort handler that can be used to respond, but when a program crashes in other ways such as a JS/WASM exception there's currently no good way to handle this consistently.

Exceptions are caught and dealt with when running main (in `callMain()`) and the program passes the exception to quit_, which can be overridden in a custom shell. So that's pretty much OK, although I'd prefer a way where we don't have to use a custom shell, but that's nitpicking.

Moving on, the main loop runner also catches the exception, logs it via `err `and then rethrows it so it ends up as an uncaught exception in the browser. Currently the only way to intercept these seem to be either checking the `err` logger if it starts with 'exception thrown' (yuck!) or registering an `error` event handler on the window and intercepting it there, but that will intercept all uncaught exceptions thrown on the page and filtering them is not that straightforward.

Finally, when external code calls into Emscripten there is no try/catch anywhere. This is true for cwrap, ccall and embind. So now we either need the caller to handle the exceptions or we need to proxy each and every exposed function through a JS wrapper that wraps it in a try/catch block. 

In our case we're shipping a 3D engine that gets embedded into online apps and we really don't want to burden our users with manually handling the various crashes that can happen due to bugs in our code. What we need is a common way for us to handle them and simply let them know we've crashed and pass the exception to a handler they register, then shut everything down nicely so that we don't keep crashing repeatedly after we've crashed already (crashception yay!). We need to stop the main loop, shutdown audio and GL, remove any event listeners that could pause/resume the main loop or otherwise cause our code to get invoked as it's all left in an undefined state once it's crashed. 

Maybe I've missed something, but there doesn't seem to be a mechanism like this available and it seems like it would be fairly straightforward to add something like `Module.onError/onException` or even just make sure that all cases call the shell `quit_` method. 

In general it feels like there should be a mechanism to do this similar to how native platforms usually have a way to register an unhandled exception handler for a process. And in this case, the process is the Emscripten app, not the whole page.

I can put up a PR for this if it's something that makes sense, but I'm also open to anyone telling me "this hasn't been done because reason XY, so just leave it" before I spend too much time on it :)

**TL;DR: Is there a way to easily catch any unhandled exceptions that occur within an Emscripten app and not the whole page?**

**Comment by sbc100** [comment]
You made me read the whole thing by putting the TL;DR at the end :)

I guess this seems like a reasonable request.  Are there existing conventions from the JS ecosystem can we should be following?    What do JS modules/libraries do to make this easy?  e.g. if I use something like react, or 3js, what happens when they crash?  I was assuming that its up to the whole page to handle exceptions at the outer layer rather than each library or module handling its own internal crashes.    But I know very little of the JS world...

**Comment by kripken** [comment]
I think we haven't added any special API for this in emscripten, but the web platform already has a way to do it, by listening to the "error" event,
```html
<html>
<body>
<script>

addEventListener("error", (error) => {
  console.error("An error occurred:", error);
});

throw "oh no";

</script>
</body>
</html>
```

That adds a listener, and throws an example error which is then noticed by the handler, in which you can run some code.

If I understood the question correctly and this addresses the issue, then it might make sense to document this as it's not the first time I've heard the question. I think a PR with docs would be great!

**Comment by sbc100** [comment]
I think the OP's question specifically related to *not* wanting to catch errors for the whole page:  "Is there a way to easily catch any unhandled exceptions that occur within an Emscripten app and not the whole page?"

**Comment by kripken** [comment]
Oh sorry, I missed that.

Is filtering by the error message an option? A wasm exception will start with "RuntimeError", for example.

Another option is to intervene in the middle, e.g. replace main (or the main loop function) with this:
```cpp
EMSCRIPTEN_KEEPALIVE
int real_main() {
  ..
}

int main() {
  return EM_ASM_INT({
    try {
      return _real_main();
    } catch (e) { ..handle it.. }
  });
}
```

**Comment by tklajnscek** [comment]
**TL;DR (@sbc100 😅) - I tried this and it has issues. I'll get a PR up for review and you guys can see if it makes sense or not.**

First of all, sorry for the late reply - I built Emscripten/Clang from source yesterday to start working on the PR for this and hit a compiler bug in Clang so I spent the day figuring that out and filing bug reports. Gotta love Mondays 🤪 

Anyhow, the first thing I tried was filtering the errors by their type, but it's really not that straightforward, as there are quite a few different ones that can get thrown. It also fails once you have another thing on the page that can throw the same error type (e.g. another WASM program). This is also further complicated by the fact that our code calls JS callbacks that can also throw an exception which is obviously not our problem, but if uncaught it still leaves our code in an undefined state due to the JS task bailing out. At this point it get's almost impossible to filter so I dropped the idea.

Then I tried wrapping the functions in try/catch block, but it got really bloated really fast - the main and mainloop functions are easy enough to do because there's just two of them. But we also expose ~100 or so API calls (via embind) and the code running within them can fail too. And those are not that straightforward to hook in bulk. If I want to keep the try/catch in our code I need to wrap each cpp function as they each have different signatures so I can't hook it in bulk (I spent a bunch of time trying some nasty template magic, but we're still targeting C++11 so we're a bit limited with parameter packs and such).

So in the end, I thought I'd ask in case I missed something obvious, which doesn't seem to be the case. I'll put up a PR soon just to show what it would take and then you guys can decide if it's worth including or not. The changes actually ended up being very minimal.

**Comment by tklajnscek** [comment]
Oh and there's a further point I forgot about: Would it make sense to bail out of exported function calls the same way as the mainloop and some others bail out if the code has ABORTed? It could easily be done in the invoker functions.

Otherwise we might have the user calling into the code that's in a broken state causing cascading failures that are harder to reason about.

I know it's easy to say _they shouldn't do it after it crashes_ but we all know it's non-trivial to ensure that - there's a bunch of stuff in flight usually, things scheduled, promises pending etc.

**Comment by sbc100** [comment]
I'm trying to understand how this issue is emscripten specific.   

Its this a problem for all JS code on the web.   If you have a web page that uses 100 different JS libraries and one of them crashes and goes into a bad state, its likely that other things could still call those functions and get cascading errors, right?   Is there some reason why emscripten-generated libraries are any different here?   What am I missing?   It it just the they tend to get into a bad state more easily?

**Comment by sbc100** [comment]
@RReverser who know way more about the wider web stuff than I do.

**Comment by kripken** [comment]
> But we also expose ~100 or so API calls (via embind) and the code running within them can fail too. And those are not that straightforward to hook in bulk.

How about a loop over all the exports to wrap them automatically?

```
for (var ex : Module) {
  if (typeof ex === 'function') {
    var original = Module[x]; // XXX will need a closure here due to JS loop semantics
    Module[ex] = function() {
      try { return original.apply(null, arguments) } catch { .. }
    };
  }
}
```

But this is a good point:

> Would it make sense to bail out of exported function calls the same way as the mainloop and some others bail out if the code has ABORTed?

Yes, I think that would make sense to do. We should already check it in the main loop, but may be missing other entry points.

**Comment by RReverser** [comment]
So, as far as I understand, the problem here is that Emscripten invokes main asynchronously, and then an exception might be thrown asynchronously too, without a good way to "catch" it without global handlers. Is that understanding of the issue correct?

If so, then I think the right way to handle this is either via `Promise` integration (I think the API is called `ready()` now?) or via `noInitialRun`, so that you wait for the module to be fully loaded, and then can put a regular JS `try-catch` around the `main` invocation and catch any errors inside.

Does this make sense?

**Comment by tklajnscek** [comment]
> I'm trying to understand how this issue is emscripten specific.
> 
> Isn't this a problem for all JS code on the web. If you have a web page that uses 100 different JS libraries and one of them crashes and goes into a bad state, its likely that other things could still call those functions and get cascading errors, right? Is there some reason why emscripten-generated libraries are any different here? What am I missing? It it just the they tend to get into a bad state more easily?

I guess there isn't _really_ a huge difference, but if I was doing a plain JS library it would be fairly straightforward to manage this the way I want because I would be writing the whole thing in a JavaScripty way. And I'd just wrap all my API calls in a simple forward-all-args wrapper function in addition to main & RAF and call it a day.

With our native app compiling with Emscripten to JS/WASM we're losing a bunch of that control (for good reason, I ain't complaining!). There are layers of glue code to make it all work nicely and we have to start doing hacky things to handle this without manually instrumenting all possible entrypoints into native code and the embind setup becomes a huge mess.

Furthermore, I think a JS lib that threw an assertion could still keep nicely executing other things after that because the VM takes care of cleanup whereas an Emscripten program is emulating a bunch of native things that will be left dangling (destructors will not have been called, the stack will have been left wherever it was when we threw etc). This would be fine if we were just a regular program - we'd crash and get terminated by the OS. But when an Emscripten app crashes it just breaks out of the current JS task and there's nothing stopping it getting called again. Well, unless it aborts, then there are some protections to handle this - I just feel like they could be a bit broader.

Finally, when we crash we'd really want to know we have crashed as we want to stop a bunch of things that are running in the page context and may cause artifacts when the app crashes (e.g. there's audio output that can start stuttering/looping garbage) and we need to free as many resources as possible (e.g. lose the GL context etc as we start hitting weird instance count limits during development when a single tab is reused all the time). 

So this affects multiple things:
- the end user experience of a crash and our ability to tell them what has happened, because we could shutdown nicely and not draw garbage or play erroneous audio
- the developer (our user) experience because we make it very clear to them when stuff has gone wrong in our code vs something else
- the error reporting and logging, because it stops error-ception from happening once we've crashed and still get called into (this is also the thing that prompted this post - I spent so much time trying to figure out what went wrong and then realized we've crashed a few seconds ago 🤯)

Right now I have a hacky workaround (by copy/pasting embind's internal templates and adding our hooks into them) which lets us handle this nicely. We cause the program to abort on an unhandled exception and then throw in all embind calls if `ABORT == true` with a clear 'You're trying to use the player after it has crashed' message. But it's really ugly the way it's done and it's going to need re-copying those templates everytime we update emscripten so it's not a workable solution by a long shot.

> How about a loop over all the exports to wrap them automatically?

Haha, I keep forgetting this is JavaScript 🤣. That might work, I'll try it out!

> Yes, I think that would make sense to do.

- main & mainloop are fine
- ccall/cwrap are easy fixes
- embind is a bit more involved because the invoker functions are generated and don't see 'var ABORT'. With assertions enabled there's another wrapper, but that's only there in assertion builds. I'll investigate.

> So, as far as I understand, the problem here is that Emscripten invokes main asynchronously, and then an exception might be thrown asynchronously too, without a good way to "catch" it without global handlers. Is that understanding of the issue correct?

main will actually invoke quit with the exception so that one is pretty much ok.
The mainloop (RAF) function is more problematic.
But any exported functions are even more problematic (I'll try the wrapping @kripken suggested above)

**Comment by tklajnscek** [comment]
> How about a loop over all the exports to wrap them automatically?

Eh @kripken , this will also:
- wrap all Emscripten's internal functions and objects (they _are_ functions in JS)
- it will wrap things multiple times (e.g. it will wrap an exposed func but also the dynCall_xxx invoker). 
- wrap any user-registered things (e.g. onAbort etc)

I could do some fancy filtering on the names of the symbols, but I already see the ones that will fall through the cracks because someone won't know they have to update a filter somehwere :)

**Comment by tklajnscek** [comment]
Put up a PR for reference here:
https://github.com/emscripten-core/emscripten/pull/11690

The changes are fairly minimal. This only addresses the notifying part of the problem, not the bailing out of calls into aborted program.

**Comment by tklajnscek** [comment]
> > Would it make sense to bail out of exported function calls the same way as the mainloop and some others bail out if the code has ABORTed?
> 
> Yes, I think that would make sense to do. We should already check it in the main loop, but may be missing other entry points. edit: but we'd need to consider code size costs here too. At minimum, though, we should do it in ASSERTIONS mode if we don't already.

Added this in https://github.com/emscripten-core/emscripten/pull/11694
